src/App.ts

```diff
@@ -795,174 +760,173 @@ const validViewTypes = ['view_closed', 'view_submission'];
-          : undefined,
-      userId:
-        type === IncomingEventType.Event
-          ? typeof (body as SlackEventMiddlewareArgs['body']).event.user === 'string'
-            ? ((body as SlackEventMiddlewareArgs['body']).event.user as string)
-            : typeof (body as SlackEventMiddlewareArgs['body']).event.user === 'object'
-            ? ((body as SlackEventMiddlewareArgs['body']).event.user.id as string)
-            : (body as SlackEventMiddlewareArgs['body']).event.channel !== undefined &&
-              (body as SlackEventMiddlewareArgs['body']).event.channel.creator !== undefined
-            ? ((body as SlackEventMiddlewareArgs['body']).event.channel.creator as string)
-            : (body as SlackEventMiddlewareArgs['body']).event.subteam !== undefined &&
-              (body as SlackEventMiddlewareArgs['body']).event.subteam.created_by !== undefined
-            ? ((body as SlackEventMiddlewareArgs['body']).event.subteam.created_by as string)
-            : undefined
-          : type === IncomingEventType.Action ||
-            type === IncomingEventType.Options ||
-            type === IncomingEventType.ViewAction ||
-            type === IncomingEventType.Shortcut
-          ? ((body as (SlackActionMiddlewareArgs | SlackOptionsMiddlewareArgs | SlackViewMiddlewareArgs)['body']).user
-              .id as string)
-          : type === IncomingEventType.Command
-          ? ((body as SlackCommandMiddlewareArgs['body']).user_id as string)
-          : undefined,
-      conversationId: channelId,
-      isEnterpriseInstall: false,
-    };
+
+  const teamId: string | undefined = (() => {
+    if (type === IncomingEventType.Event) {
+      const bodyAsEvent = body as SlackEventMiddlewareArgs['body'];
+      if (
+        Array.isArray(bodyAsEvent.authorizations) &&
+        bodyAsEvent.authorizations[0] !== undefined &&
+        bodyAsEvent.authorizations[0].team_id !== null
+      ) {
+        return bodyAsEvent.authorizations[0].team_id;
+      }
+      return bodyAsEvent.team_id;
+    }
+
+    if (type === IncomingEventType.Command) {
+      return (body as SlackCommandMiddlewareArgs['body']).team_id;
+    }
+
+    if (
+      type === IncomingEventType.Action ||
+      type === IncomingEventType.Options ||
+      type === IncomingEventType.ViewAction ||
+      type === IncomingEventType.Shortcut
+    ) {
+      const bodyAsActionOrOptionsOrViewActionOrShortcut = body as (
+        | SlackActionMiddlewareArgs
+        | SlackOptionsMiddlewareArgs
+        | SlackViewMiddlewareArgs
+        | SlackShortcutMiddlewareArgs
+      )['body'];
+
+      // When the app is installed using org-wide deployment, team property will be null
+      if (bodyAsActionOrOptionsOrViewActionOrShortcut.team !== null) {
+        return bodyAsActionOrOptionsOrViewActionOrShortcut.team.id;
+      }
+
+      // This is the only place where this function might return undefined
+      return bodyAsActionOrOptionsOrViewActionOrShortcut.user.team_id;
```

Comment:
Do we ever need to worry about `bodyAsActionOrOptionsOrViewActionOrShortcut.user` being undefined? The pre refactored code had a check for this.

Reply:
Nope, as long as our type definitions are correct this should be safe. I just manually traced all the possible types for `bodyAsActionOrOptionsOrViewActionOrShortcut.user` and each of them have the property defined, non-optionally.

Gotta love that TypeScript magic ✨ .